Port-StreamBase

MakoEnergy opened the pull request:
Initial implementation of Stream base classes.

MakoEnergy committed:
Initial implementation of Stream base classes.

MakoEnergy commented:
This PR is a little weird, because the classes introduced in it lack tests.  There are a series of pure virtual base classes, but also 3 classes with partial implementations while still lacking implementations of 6 functions.  

It is possibly to jank together an implementation, but doing so would likely result in a class remarkably similar to the next classes to be ported.

MakoEnergy committed:
A round of fixes for clang.

Added warning supression for padding.  Added explicit casts for the signed to unsigned conversions.  Touched up some of the docs.

MakoEnergy committed:
Another round of fixes.

More of the same as the last commit, plus created some missing implementations.

MakoEnergy committed:
Added a warning supression for weak vtables.

Clang doesn't like that we don't have a source file with even one definition for the pure abstract classes that are the stream base interfaces.  So it doesn't know where to put the vtables for those classes...but it'll manage.

MakoEnergy committed:
Fixed a missing cast.

MakoEnergy committed:
Fixed a memory layout issue is MSVC.

Sqeaky commented on include/IOStream.h:
```diff
@@ -0,0 +1,136 @@
+    {
+    public:
+        /// @brief Class constructor.
+        /// @param Buf A pointer to the buffer that will be streamed.
```
Does this take ownership of the pointer?

MakoEnergy replied:
I don't believe it does.  The buffer is passed directly into the standard iostream.  Generally the expectation is that the buffer is a data member of the derived class so cleanup is automatic.  Or at least that's what most implementations I've seen do.

Sqeaky commented on include/StreamBase.h:
```diff
@@ -0,0 +1,230 @@
+    };//iStreamBase
+
+    #ifdef _MSC_VER
+    #pragma vtordisp(push, 2)
```
What is this? A comment is probably reasonable.

MakoEnergy replied:
A comment was added.

Sqeaky commented on include/StreamBase.h:
```diff
@@ -0,0 +1,230 @@
+        /// @param Buffer The buffer to be populated with the read data.
+        /// @param Size The number of bytes to read from the Stream.
+        /// @return Returns the number of bytes successfully read.
+        virtual size_t Read(void* Buffer, const StreamSize Size) = 0;
```
These kinds of APIs always rub me the wrong the wrong way. They are error prone and not easily composable. I am glad there is a GetAsString() member that always this to be used without the intentional passing in of a mutable item. I know this is one of the higher performance options, but there are ways to keep that performance that are at least less error prone, like the BinaryBuffer.

We should discuss alternatives. I know this is a point of contention we often have.

Sqeaky replied:
We discussed this. For others reading this, my takeaway was that there would be simplification here, removing context specific functions. These helper functions would move to additional classes like TextStreamReader and BinaryStreamReader classes that would have context specific helper functions.

Stuff like ReadLine() and GetAsString() could go onto the TextStreamReader.

Sqeaky commented on src/IOStream.cpp:
```diff
@@ -0,0 +1,164 @@
+        String Ret;
+        size_t RetSize = static_cast<size_t>( this->GetSize() );
+        if( RetSize > 0 ) {
+            Ret.reserve(RetSize);
```
+1 for the null terminator?

MakoEnergy replied:
https://stackoverflow.com/questions/30111288/stdstringreserve-and-end-of-string-0

MakoEnergy committed:
Added a series of Binary/Test Readers/Writers.

Also touched up small parts of the IOStream API to make them slightly cleaner.

MakoEnergy committed:
More code cleanup for Clang

Changed SizeType to StreamSize where applicable.  Added a couple exceptions.  Added a throw test.

MakoEnergy committed:
Two more casts because Clang is picky.

MakoEnergy committed:
Use numeric_limits instead of tricks.

MakoEnergy committed:
Another round of casts into SizeType.

MakoEnergy committed:
More casts, this time in TextStreamReader.

MakoEnergy committed:
Suppress dumb clang warning.

MakoEnergy committed:
Emscripten is last to go.

MakoEnergy committed:
Hopeful fix for *raspian.  Not emscripten.

Sqeaky commented on include/InputOutputStream.h:
```diff
@@ -0,0 +1,134 @@
+// � Copyright 2010 - 2019 BlackTopp Studios Inc.
```
Is this the copyright symbol?

MakoEnergy replied:
Yes.

Sqeaky commented on include/TextStreamReader.h:
```diff
@@ -0,0 +1,104 @@
+        /// then set to the beginning of the Stream, advanced to the end, and then set back to the same read
+        /// position the Stream was at when this function was called.
+        /// @n @n
+        /// This might produce a very large String. Be mindful of the Streams size when using this.
```
You suggest being mindful of the stream size here. Yet there is no function for gathering that information. In other places you noted when you expected the game dev to know things. Should the game dev come with a good idea. should there be a size function on this, or would it be prudent to @ref some function on the InputStream?

Also, per our discussion some streams are impossible to check the size of. Is it possible to make a function to check if the stream is can be reasonably read at once?

MakoEnergy replied:
The check already exists.  In the third line of the function where we check the value of the returned result of tellg, we're seeing if we got a gibberish value.  If we simply cannot seek in the stream, we don't bother and immediately return an empty String.

Alternatively, the developer can check the stream before creating a Reader with it.  If it's just a standard stream then they can try to seek with it themselves, or they may also have an idea of where it came from because presumably they initialized it.  If it's one of our streams then they can query the size of the stream directly with our API.

Sqeaky commented on src/BinaryStreamReader.cpp:
```diff
@@ -0,0 +1,80 @@
+
+    BinaryBuffer BinaryStreamReader::Read(const StreamSize Bytes)
+    {
+        BinaryBuffer::Byte* MaybeReturn = new BinaryBuffer::Byte[static_cast<SizeType>(Bytes)];
```
This should throw std::bad_alloc when it fails. This could be noted in the documentation.

Sqeaky commented on src/TextStreamReader.cpp:
```diff
@@ -0,0 +1,112 @@
+    String TextStreamReader::Read(const StreamSize Bytes)
+    {
+        String ToReturn;
+        ToReturn.resize(static_cast<SizeType>(Bytes),'\0');
```
There is a constructor that can accept a size and an initializer char.  The 2nd constructor on this page: https://en.cppreference.com/w/cpp/string/basic_string/basic_string

MakoEnergy replied:
Fixed.

MakoEnergy committed:
Minor code cleanup.
